fuel-p2p/src/service.rs

```diff
@@ -554,6 +553,135 @@ mod tests {
         }
     }
 
+    // Simulate 2 Sets of Sentry nodes.
+    // In both Sets, a single Guarded Node should only be connected to their sentry nodes.
+    // While other nodes can and should connect to nodes outside of the Sentry Set.
+    #[tokio::test]
+    #[instrument]
+    async fn sentry_nodes_working() {
+        let reserved_nodes_size = 4;
+        let double_reserved_nodes_size = reserved_nodes_size * 2;
+
+        let mut p2p_config = P2PConfig::default_with_network("sentry_nodes_working");
+        // enable mdns for faster discovery of nodes
+        p2p_config.enable_mdns = true;
+
+        // The desired mesh number is double of the size of the reserved nodes.
+        // This way the non-guarded nodes will want to try to connect to as many nodes as possible in the network.
+        // This still should not affect the guarded nodes that are enabling `reserved_nodes_only_mode`.
+        let gossipsub_config = default_gossipsub_builder()
+            .mesh_n(double_reserved_nodes_size)
+            .mesh_n_low(double_reserved_nodes_size)
+            .mesh_n_high(double_reserved_nodes_size)
+            .build()
+            .expect("valid gossipsub configuration");
+
+        p2p_config.gossipsub_config = gossipsub_config;
+
+        let build_sentry_nodes = || async {
+            let instantiate_nodes: Vec<_> = (0..reserved_nodes_size)
+                .map(|_| async {
+                    let mut node = build_fuel_p2p_service(p2p_config.clone());
```

Comment:
We should also set the guarded node as reserved for the sentries, so that the don't accept too many other random peers and drop their connection to the guarded node they're assigned to.

Reply:
In order to achieve this, I needed to refactor a bit the way I initialize p2p services within tests, previously I didn't know a nodes' address until it started,  only then I could pass its address to another node's bootstrap/reserved nodes list.

And since here we wanted that both sentry nodes and guarded nodes have each other address specified on start, we had a chicken and egg problem.

I created `NodeData` struct to deal with this issue, and consequently I refactored other tests to use this approach - it was a bit cleaner instead of .awaiting for node's address every time.